cache/clock_cache.cc

```diff
@@ -238,45 +282,60 @@ void ClockCacheShard::ApplyToSomeEntries(
         callback(h->key(), h->value, h->GetCharge(metadata_charge_policy),
                  h->deleter);
       },
-      index_begin, index_end);
+      index_begin, index_end, false);
 }
 
 void ClockCacheShard::ClockRemove(ClockHandle* h) {
-  assert(h->IsInClockList());
   h->SetClockPriority(ClockHandle::ClockPriority::NONE);
-  assert(clock_usage_ >= h->total_charge);
-  clock_usage_ -= h->total_charge;
 }
 
 void ClockCacheShard::ClockInsert(ClockHandle* h) {
-  assert(!h->IsInClockList());
+  assert(!h->IsInClock());
   bool is_high_priority =
       h->HasHit() || h->GetCachePriority() == Cache::Priority::HIGH;
   h->SetClockPriority(static_cast<ClockHandle::ClockPriority>(
       is_high_priority * ClockHandle::ClockPriority::HIGH +
       (1 - is_high_priority) * ClockHandle::ClockPriority::MEDIUM));
-  clock_usage_ += h->total_charge;
+}
+
+void ClockCacheShard::Evict(ClockHandle* h) {
+  ClockRemove(h);
+  table_.Remove(h);
+  assert(usage_ >= h->total_charge);
+  usage_ -= h->total_charge;
 }
 
 void ClockCacheShard::EvictFromClock(size_t charge,
                                      autovector<ClockHandle>* deleted) {
+  // TODO(Guido) When an element is in the probe sequence of a
+  // hot element, it will be hard to get an exclusive ref.
+  // We may need a mechanism to avoid that an element sits forever
+  // in cache waiting to be evicted.
   assert(charge <= capacity_);
-  while (clock_usage_ > 0 && (usage_ + charge) > capacity_) {
-    ClockHandle* old = &table_.array_[clock_pointer_];
+  uint32_t max_iterations = table_.GetTableSize();
+  while (usage_ + charge > capacity_ && max_iterations--) {
+    ClockHandle* h = &table_.array_[clock_pointer_];
     clock_pointer_ = table_.ModTableSize(clock_pointer_ + 1);
-    // Clock list contains only elements which can be evicted.
-    if (!old->IsInClockList()) {
-      continue;
-    }
-    if (old->GetClockPriority() == ClockHandle::ClockPriority::LOW) {
-      ClockRemove(old);
-      table_.Remove(old);
-      assert(usage_ >= old->total_charge);
-      usage_ -= old->total_charge;
-      deleted->push_back(*old);
-      return;
+
+    if (h->TryExclusiveRef()) {
+      if (!h->IsInClock() && h->IsElement()) {
+        // We re-insert the element into clock.
```

Comment:
"re-insert the element into clock" = SetClockPriority ?
So, it is not inserting into a table/array?

Reply:
I clarified the comment.

Reply:
This ties back to your previous comment about ClockInsert and ClockRemove. The comment I added there also helps here.

Reply:
I am not a fan of this terminology either

Reply:
Changed ClockInsert/ClockRemove to ClockOn/ClockOff.